File: Delphi/GildedRose.pas
unit GildedRose;

interface

uses
  Item,
  System.Generics.Collections;

type
  TGildedRose = class(TObject)
  private
    FItems: TObjectList<TItem>;
  public
    constructor Create(const AItems: TObjectList<TItem>);
    procedure UpdateQuality;
    property Items: TObjectList<TItem> read FItems;
  end;

implementation

{ TGildedRose }

constructor TGildedRose.Create(const AItems: TObjectList<TItem>);
begin
  inherited Create;
  FItems := AItems;
end;

procedure TGildedRose.UpdateQuality;
var
  I: Integer;
begin
  for I := 0 to Items.Count - 1 do
  begin
    if (Items[I].Name <> 'Aged Brie') and (Items[I].Name <> 'Backstage passes to a TAFKAL80ETC concert') then
    begin
      if Items[I].Quality > 0 then
      begin
        if Items[I].Name <> 'Sulfuras, Hand of Ragnaros' then
        begin
          Items[I].Quality := Items[I].Quality - 1;
        end;
      end;
    end
    else
    begin
      if Items[I].Quality < 50 then
      begin
        Items[I].Quality := Items[I].Quality + 1;
        if Items[I].Name = 'Backstage passes to a TAFKAL80ETC concert' then
        begin
          if Items[I].SellIn < 11 then
          begin
            if Items[I].Quality < 50 then
            begin
              Items[I].Quality := Items[I].Quality + 1;
            end;
          end;

          if Items[I].SellIn < 6 then
          begin
            if Items[I].Quality < 50 then
            begin
              Items[I].Quality := Items[I].Quality + 1;
            end;
          end;
        end;
      end;
    end;

    if Items[I].Name <> 'Sulfuras, Hand of Ragnaros' then
    begin
      Items[I].SellIn := Items[I].SellIn - 1;
    end;

    if Items[I].SellIn < 0 then
    begin
      if Items[I].Name <> 'Aged Brie' then
      begin
        if Items[I].Name <> 'Backstage passes to a TAFKAL80ETC concert' then
        begin
          if Items[I].Quality > 0 then
          begin
            if Items[I].Name <> 'Sulfuras, Hand of Ragnaros' then
            begin
              Items[I].Quality := Items[I].Quality - 1;
            end;
          end;
        end
        else
        begin
          Items[I].Quality := Items[I].Quality - Items[I].Quality;
        end;
      end
      else
      begin
        if Items[I].Quality < 50 then
        begin
          Items[I].Quality := Items[I].Quality + 1;
        end;
      end;
    end;
  end;
end;

end.

File: python/tests/gilded_rose.py
# -*- coding: utf-8 -*-

class GildedRose(object):
    BACKSTAGE = 'Backstage passes to a TAFKAL80ETC concert'
    SULFURAS = 'Sulfuras, Hand of Ragnaros'
    AGED_BRIE = 'Aged Brie'
    MAX_QUAL = 50

    def __init__(self, items):
        self.items = items

    def __qual_inc(self, item, inc):
        item.quality = min(self.MAX_QUAL, item.quality + inc)

    def __qual_dec(self, item, dec):
        item.quality = max(0, item.quality - dec)

    def update_quality(self):
        for item in self.items:
            match item.name:
                case self.AGED_BRIE:
                    if item.sell_in > 0:
                        self.__qual_inc(item, 1)
                    else:
                        self.__qual_inc(item, 2)
                case self.SULFURAS:
                    pass
                case self.BACKSTAGE:
                    if item.sell_in > 10:
                        self.__qual_inc(item, 1)
                    elif item.sell_in > 5:
                        self.__qual_inc(item, 2)
                    elif item.sell_in > 0:
                        self.__qual_inc(item, 3)
                    else:
                        item.quality = 0
                case _:
                    multiplier = 1
                    if item.name.startswith('Conjured'):
                        multiplier = 2
                    if item.sell_in > 0:
                        self.__qual_dec(item, multiplier)
                    else:
                        self.__qual_dec(item, 2 * multiplier)
            if not item.name == self.SULFURAS:
                item.sell_in -= 1


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)
